### PR TITLE
Fix warnings when using cmake 3.18

### LINKED ISSUE #10756: CMake 3.18: warnings when configuring with CUDA
We get a bunch of warnings with CMake 3.18. Below is the new policy that causes the warnings

```cmake --help-policy CMP0104
CMP0104
-------

Initialize ``CMAKE_CUDA_ARCHITECTURES`` when
``CMAKE_CUDA_COMPILER_ID`` is ``NVIDIA``.
Raise an error if ``CUDA_ARCHITECTURES`` is empty.

``CMAKE_CUDA_ARCHITECTURES`` introduced in CMake 3.18 is used to
initialize ``CUDA_ARCHITECTURES``, which passes correct code generation
flags to the CUDA compiler.

Previous to this users had to manually specify the code generation flags. This
policy is for backwards compatibility with manually specifying code generation
flags.

The ``OLD`` behavior for this policy is to not initialize
``CMAKE_CUDA_ARCHITECTURES`` when
``CMAKE_CUDA_COMPILER_ID`` is ``NVIDIA``.
Empty ``CUDA_ARCHITECTURES`` is allowed.

The ``NEW`` behavior of this policy is to initialize
``CMAKE_CUDA_ARCHITECTURES`` when
``CMAKE_CUDA_COMPILER_ID`` is ``NVIDIA``
and raise an error if ``CUDA_ARCHITECTURES`` is empty during generation.

If ``CUDA_ARCHITECTURES`` is set to a false value no architectures
flags are passed to the compiler. This is intended to support packagers and
the rare cases where full control over the passed flags is required.

This policy was introduced in CMake version 3.18.  CMake version
3.18.0 warns when the policy is not set and uses ``OLD`` behavior.
Use the ``cmake_policy()`` command to set it to ``OLD`` or ``NEW``
explicitly.

.. note::
  The ``OLD`` behavior of a policy is
  ``deprecated by definition``
  and may be removed in a future version of CMake.

Examples
^^^^^^^^

 set_property(TARGET tgt PROPERTY CUDA_ARCHITECTURES 35 50 72)

Generates code for real and virtual architectures ``30``, ``50`` and ``72``.

 set_property(TARGET tgt PROPERTY CUDA_ARCHITECTURES 70-real 72-virtual)

Generates code for real architecture ``70`` and virtual architecture ``72``.

 set_property(TARGET tgt PROPERTY CUDA_ARCHITECTURES OFF)

CMake will not pass any architecture flags to the compiler.
```

### PR DESCRIPTION
**Pull request by Rombur**
This fixes the warnings I get with cmake 3.18 and CUDA. 

Closes #10756

### CODE REVIEW

File: examples/CMakeLists.txt [hunk 1]
```diff
@@ -112,6 +112,10 @@ IF(DEAL_II_COMPONENT_EXAMPLES)
               )
           ENDIF()
 
+          IF(CMAKE_VERSION VERSION_GREATER_EQUAL 3.18)
```
Comment (tjhei) [inline examples/CMakeLists.txt:115]: Why do you need to do this here and not inside the individual examples? Is it possible to put this inside one of the setup_target cmake macros?

**Comment by tamiko** [comment]
@Rombur Would you mind testing whether the following alternative works?
```
diff --git a/cmake/config/Config.cmake.in b/cmake/config/Config.cmake.in
index e9323ffe48..af42960d37 100644
--- a/cmake/config/Config.cmake.in
+++ b/cmake/config/Config.cmake.in
@@ -149,6 +149,11 @@ SET(DEAL_II_MPIEXEC_NUMPROC_FLAG "@MPIEXEC_NUMPROC_FLAG@")
 SET(DEAL_II_MPIEXEC_PREFLAGS "@MPIEXEC_PREFLAGS@")
 SET(DEAL_II_MPIEXEC_POSTFLAGS "@MPIEXEC_POSTFLAGS@")
 
+#
+# CUDA specific setup:
+#
+SET(CMAKE_CUDA_ARCHITECTURES "@CMAKE_CUDA_ARCHITECTURES@")
+
 #
 # Build a static executable:
 #
diff --git a/cmake/configure/configure_1_cuda.cmake b/cmake/configure/configure_1_cuda.cmake
index 3dcff92305..8a5e65b8d7 100644
--- a/cmake/configure/configure_1_cuda.cmake
+++ b/cmake/configure/configure_1_cuda.cmake
@@ -173,6 +173,14 @@ MACRO(FEATURE_CUDA_CONFIGURE_EXTERNAL)
   SET(CMAKE_CUDA_USE_RESPONSE_FILE_FOR_LIBRARIES 0)
   SET(CMAKE_CUDA_USE_RESPONSE_FILE_FOR_OBJECTS 0)
 
+  #
+  # Disable CUDA_ARCHITECTURES target properties
+  #
+  IF(CMAKE_VERSION VERSION_GREATER_EQUAL 3.18)
+    SET(CMAKE_CUDA_ARCHITECTURES OFF)
+  ENDIF()
+
+
   #
   # Set up cuda flags:
   #
```

**Comment by Rombur** [comment]
@tamiko This works for the library but I still get warnings for the testsuite.

**Comment by Rombur** [comment]
I actually messed up when I applied @tamiko's patch. It works fine!

**Review by masterleinad — APPROVED** [review]

**Pull request merged**